### ISSUE TITLE
[$500] Feature Request: Add full screen expand button when editing a messge

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15596]
Held on https://github.com/Expensify/App/issues/16078
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Problem:
When there's multiple lines being edited, there is a "full-screen expand" button in the upper-left of the compose window, but not when editing an existin

## Solution:
Add the full-screen expand button to the upper-left when editing an existing comment

## Context/Examples/Screenshots/Notes:

**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/222207842-4cb95747-e11d-4ab7-a3c9-6a3f85cdedd9.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1677476774053969

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e878e8c8ed91741e</li>
        <li>Upwork Job ID: 1632674482135707648</li>
        <li>Last Price Increase: 2024-11-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @s77rt</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sonialiap (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by sonialiap** [comment]
I can't think of a reason for why we wouldn't want to allow expanding the editor to full screen 👍

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cristipaval (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by cristipaval** [comment]
This could be external.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01e878e8c8ed91741e

**Comment by MelvinBot** [comment]
Current assignee @sonialiap is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by MelvinBot** [comment]
Current assignee @cristipaval is eligible for the External assigner, not assigning anyone new.

**Comment by tienifr** [comment]
Since this is similar to the existing `Composer` and is quite straight-forward,     I can take it if it's on volunteer-basis.

**Comment by s77rt** [comment]
@tienifr I believe this still requires proposals to follow the template.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When there's multiple lines being edited, there is a "full-screen expand" button in the upper-left of the compose window, but not when editing an existing.

### What is the root cause of that problem?
This is because we don't have that feature yet when editing the message.

### What changes do you think we should make in order to solve the problem?

We can add this component https://github.com/Expensify/App/blob/f2bded26692e7b462c758241860e351e524af730/src/pages/home/report/ReportActionCompose.js#L519 into the `src/pages/home/report/ReportActionItemMessageEdit.js`

### What alternative solutions did you explore? (Optional)
NA

**Comment by s77rt** [comment]
Thanks for the proposal @tienifr. One thing I'm concerned about is how will you make the composer take full height being a FlatList item.

**Comment by tienifr** [comment]
@s77rt we can do that by making the height of the `ReportMessageItem` equal to the computed height of the Report Dropzone https://github.com/Expensify/App/commit/9591146685c52e0b00bc9fdeef4aa5ee61bafe61, if we want to keep the main composer visible then we'll minus the height of it from the Report Dropzone height.

Here's the demo:
<img width="983" alt="Screenshot 2023-03-06 at 17 53 32" src="https://user-images.githubusercontent.com/113963320/223090546-ed8c0f05-85b0-492e-8222-986c60f1fad7.png">

**Comment by s77rt** [comment]
@tienifr Thanks for the quick reply. I still need to get aligned with you on some points:
1. Will we still be able to scroll the messages? I think we should prevent that
2. Can we adjust the height without relaying on another component. Dropzone have nothing to do with message edit here.

**Comment by tienifr** [comment]
@s77rt thanks, your mentioned points can be addressed by: If there's a current full-size report action component, we'll return only 1 `ReportActionItem` and do not return the full `ReportActionsList`, the layers of containers of  `ReportActionItem` will be set to have `flex: 1` so it will occupy the full screen height, similar to the current behavior of `ReportActionCompose` when it has full screen height

**Comment by Nayan120603** [comment]
i read your problem of expanding button feature. here i wrote some codes for it. may be it will helpful..

here it is  https://1drv.ms/w/s!Amsi9ANVEhq5ghTx-bZVbg7Y95Gt?e=XBuWZf

**Comment by MelvinBot** [comment]
📣 @Nayan120603! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by s77rt** [comment]
@tienifr Thanks for the clarification. I think your proposal is getting a little bit complex, let me get a confirmation for the expected behaviour first.

@cristipaval @sonialiap Can you confirm if the expected behaviour is to show the edit composer full height and prevent scrolling to other messages while in this state? Or the rest of message should still be visible and reachable.

**Comment by s77rt** [comment]
@Nayan120603 Thanks for your interest. Please read the [contributing guide](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) and follow the proposal template.

**Comment by cristipaval** [comment]
> @cristipaval @sonialiap Can you confirm if the expected behaviour is to show the edit composer full height and prevent scrolling to other messages while in this state? Or the rest of message should still be visible and reachable.

I think we should be consistent and make it behave the same as when creating a new message. If you read the Slack conversation, the reporter was expecting to see the same behaviour.

**Comment by s77rt** [comment]
@cristipaval Thanks! I have asked on that slack thread.

**Comment by s77rt** [comment]
@tienifr Can you update your proposal given that we would keep the report actions items untouched (they won't be visible since the report action being edited should take full height) but we should prevent flat list from scrolling.
1. How would you make the report action take full height
2. How will you prevent scrolling of the flatlist

**Comment by tienifr** [comment]
@s77rt what do you think about this? https://github.com/Expensify/App/issues/15596#issuecomment-1455980766

It will work for both the points you mentioned. Having a not-scrollable flat list with only 1 item visible is the same as showing only that 1 item.

**Comment by s77rt** [comment]
@tienifr The additional logic of rendering only one item although it may work it seems unnecessary, it would be better to just work on top of what we have, set the height of the report action to take full height and prevent scrolling.

**Comment by tienifr** [comment]
@s77rt sure, if we want to keep the report actions list then we can do this:

> 1. How would you make the report action take full height

If an edit composer is currently full height, we should make the `FlatList` and all levels of containers of the Edit composer to have `flex: 1`. This will make sure the Action Item that has the edit composer will stretch to have full height.

> 2. How will you prevent scrolling of the flatlist

We need to apply `scrollEnabled: false` to the Flatlist (which will work for native) and `overflow: hidden` (which will work for web, we might consider adding an upstream fix to `react-native-web` since the  `scrollEnabled: false`  property is not working there)

**Comment by s77rt** [comment]
@tienifr Thanks for the quick update, can you elaborate where to put overflow: hidden? I can still scroll the flat list.

**Comment by tienifr** [comment]
@s77rt it should be added to the `contentContainerStyle` here https://github.com/Expensify/App/blob/2c35e7abfeab34b3e1ee03207d93ef7984f80a52/src/pages/home/report/ReportActionsList.js#L183
Make sure to have both these properties: `{overflow: 'hidden', flex: 1},`

**Comment by s77rt** [comment]
@tienifr Thanks once again for the clarification. I think the provided info are enough to start working on a PR, other details will be discussed there.

:ribbon: :eyes: :ribbon: C+ reviewed

cc @cristipaval

**Comment by tienifr** [comment]
thanks @s77rt for the review, since it's a feature request should we have design's confirmation on the UI (things like location of the expand button, how full-page editing would look) before moving forward?

We have 2 UI cases here:
1. The expand button on the MessageEdit
<img width="1137" alt="Screenshot 2023-03-06 at 16 44 54" src="https://user-images.githubusercontent.com/113963320/223074445-030a260e-dd26-438f-a640-b3a628df225a.png">

2. The full-screen edit after clicking on the expand button
<img width="975" alt="Screenshot 2023-03-07 at 22 57 35" src="https://user-images.githubusercontent.com/113963320/223476907-0ed9635e-01af-46fe-bb5e-b1dc87b938d7.png">

**Comment by s77rt** [comment]
@tienifr The layout should be the same as that of the main composer, the expand button should appear on the left.

**Comment by cristipaval** [comment]
Thanks @s77rt ! Looks good to me.
@tienifr , go ahead with your solution 🚀

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @cristipaval!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by cristipaval** [comment]
So sorry, @tienifr! I forgot to assign you.

**Comment by tienifr** [comment]
@s77rt  @cristipaval 

I am encountering this on the `main` branch, and I hope I could get more eyes on this. The newest code on the `main` branch, merged from [this PR](https://github.com/Expensify/App/pull/14686/files#diff-8ab8ba45ab927f1430b8d1209d187e003a9cdfe476f809e4483ef64cdba253ebL1397-L1398) has removed the full screen size for the main composer. In order for the new emoji suggestion function to work, when the user expand the main composer, it still leaves a space with report actions being displayed and scrollable.

![image](https://user-images.githubusercontent.com/113963320/224600641-3ae9ee59-4856-4edf-b459-a88a5d9d076d.png)


I think this contradicts to what we have earlier agreed upon the current feature request, so I think we need to re-evaluate on what need to be done.

1. Do we still need to make the edit composer full screen, hide other report action items and disable scrolling? Or should we make it similar to the newest main composer?
2. Do we need to make the emoji suggestion available with the edit composer?

For the sake of consistency, I think we should make it as similar as the main composer as possible. That means we would still need to display all the other messages, do not disable scrolling, and make the emoji suggestion available to the edit composer.

What do you think?

**Comment by s77rt** [comment]
1. I'm not sure, seems like a regression. cc @roryabraham 
2. For consistency I think we will but not in this issue. I assume it will be handled in another separated issue

@cristipaval thoughts?

**Comment by tienifr** [comment]
@s77rt 
I think that's not a regression. The style is intentionally modified so that the new emoji suggestions are visible. The suggestions would not be displayed if the composer height is 100%.

Either way, let's wait for @cristipaval input.

**Comment by cristipaval** [comment]
Slack discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1678802629244829)

**Comment by MelvinBot** [comment]
@s77rt @sonialiap @cristipaval @tienifr this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by s77rt** [comment]
Still waiting for a clarification on the expected behaviour. [Slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1678802629244829).

**Comment by s77rt** [comment]
@tienifr The changes are intentional, so just add the expand button and verify it has same behavior as on Main Composer. Looks like that the hard part has been cut down :tada:

**Comment by sonialiap** [comment]
Just FYI, I'll be OOO Mar 17-25

**Comment by cristipaval** [comment]
@s77rt @tienifr According to the last replies in [this](https://expensify.slack.com/archives/C035J5C9FAP/p1674741956563149) Slack convo, we might revert the changes which limited the composer height and go full space again. I think we should put this one on old until the team implements the changes su support both full screen composer and emoji suggestions on top of it.

**Comment by s77rt** [comment]
@cristipaval Makes sense. Let's hold this.

**Comment by cristipaval** [comment]
This one is held on https://github.com/Expensify/App/issues/16078

**Comment by s77rt** [comment]
On hold...

**Comment by s77rt** [comment]
Please @MelvinBot this is really on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by roryabraham** [comment]
Making this a weekly while it's on HOLD

**Comment by tienifr** [comment]
I have read and reviewed this Design Doc!

**Comment by cristipaval** [comment]
still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by cristipaval** [comment]
Still on hold.

**Comment by sonialiap** [comment]
Still on hold

**Comment by cristipaval** [comment]
still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Still on hold

**Comment by cristipaval** [comment]
Still on hold.

**Comment by sonialiap** [comment]
Still on hold

**Comment by sonialiap** [comment]
Holding pattern continues https://github.com/Expensify/App/issues/16078

**Comment by sonialiap** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by cristipaval** [comment]
Same status.

**Comment by cristipaval** [comment]
It looks like no progress yet on the issue that this one is held on.

**Comment by cristipaval** [comment]
Same status.

**Comment by s77rt** [comment]
Still on hold

**Comment by cristipaval** [comment]
same status

**Comment by s77rt** [comment]
Same ^

**Comment by cristipaval** [comment]
Still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by sonialiap** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by sonialiap** [comment]
Still on hold https://github.com/Expensify/App/issues/16078

**Comment by s77rt** [comment]
Same ^

**Comment by sonialiap** [comment]
Still on hold https://github.com/Expensify/App/issues/16078

**Comment by puneetlath** [comment]
Reopening this for proposals. Now that we have inline auto-suggestion, we should be able to make the compose box expand to take up the full available space. And we should be able to have that happen on all platforms.

**Comment by puneetlath** [comment]
Or actually, looks like @tienifr is already assigned. @s77rt @tienifr does that proposal still make sense here?

**Comment by s77rt** [comment]
@tienifr Can you please look into the part that makes the main composer full screen as well as updating your proposal to reflect the current codebase

**Comment by s77rt** [comment]
Not overdue. Waiting on proposals

**Comment by kevinksullivan** [comment]
going to take a look at this one, so I'm removing the `external` label for now.

**Comment by kevinksullivan** [comment]
Sorry, digging into this today.

**Comment by kevinksullivan** [comment]
Need to take the logic in AttachmentPickerWithMenuItems.tsx here:

https://github.com/Expensify/App/blob/b10580f1d46365025d7b7698f6c3f2e43883f7f1/src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.tsx#L259-L281

and add it to the ReportActionItemMessageEdit component

https://github.com/Expensify/App/blob/b10580f1d46365025d7b7698f6c3f2e43883f7f1/src/pages/home/report/ReportActionItemMessageEdit.tsx

**Comment by kevinksullivan** [comment]
Ok, so we have an existing function at the reportID level to determine if the expand icon should be shown, but that won't work for reportActions. 

https://github.com/Expensify/App/blob/b10580f1d46365025d7b7698f6c3f2e43883f7f1/src/libs/actions/Report.ts/#L2318-L2320

So the plan is to add a new onyx key that will support checking specific reportActionIds to inform if we show the expand icon in a message that is being edited. CC @roryabraham

**Comment by roryabraham** [comment]
I've been helping @kevinksullivan with this one, so I'm going to nab it from you @cristipaval

**Comment by cristipaval** [comment]
Thanks @roryabraham 🙌

**Comment by roryabraham** [comment]
@kevinksullivan posted in slack for consensus on the design: https://expensify.slack.com/archives/C01GTK53T8Q/p1721946623645629

**Comment by kevinksullivan** [comment]
Started working on draft PR https://github.com/Expensify/App/pull/46330 . Shoutout @s77rt for your [suggestion](https://expensify.slack.com/archives/C01GTK53T8Q/p1721959465526129?thread_ts=1721946623.645629&cid=C01GTK53T8Q) to use reportActionDrafts as well!

**Comment by s77rt** [comment]
PR still WIP

**Comment by kevinksullivan** [comment]
Same ^ should be in review next week.

**Comment by sonialiap** [comment]
@anmurali I'm OOO Aug 19-30, adding leave buddy. 
Status: PR in review

**Comment by kevinksullivan** [comment]
PR is WIP, have had travel priorities take over lately. Will aim to sync back up with Rory this week.

**Comment by kevinksullivan** [comment]
scheduled time with @roryabraham Wednesday to go over this.

**Comment by kevinksullivan** [comment]
no update right now

**Comment by kevinksullivan** [comment]
No update unfortunately, and going OOO for a bit so this will be held until then.

**Comment by kevinksullivan** [comment]
Discussing this 1:1 with Rory

**Comment by roryabraham** [comment]
Unfortunately @kevinksullivan had to drop this, but let's work to get it over the finish line. Reassigning @s77rt as C+.

**Comment by roryabraham** [comment]
adding @sonialiap back as BZ

**Comment by roryabraham** [comment]
@tienifr I think it's fair to give you first dibs on this one if you want to drop a proposal, since you worked on it before it went on HOLD for an extended period and have good context.

**Comment by s77rt** [comment]
@Shahidullah-Muffakir Thanks for the proposal. Your suggested solution is incomplete. For instance we want to save the composer full size status (`isComposerFullSize`) in onyx per report action (in `reportActionsDrafts_` collection and not `reportIsComposerFullSize_`). Also I don't think that adding `styles.chatItemFullComposeBox` style would be enough. Can you please look into this in depth and update your proposal?

**Comment by Shahidullah-Muffakir** [comment]
> @Shahidullah-Muffakir Thanks for the proposal. Your suggested solution is incomplete. For instance we want to save the composer full size status (`isComposerFullSize`) in onyx per report action (in `reportActionsDrafts_` collection and not `reportIsComposerFullSize_`). Also I don't think that adding `styles.chatItemFullComposeBox` style would be enough. Can you please look into this in depth and update your proposal?

@s77rt , Actually I didn't write all the changes needed, I will create a test branch, you can test that. Thank you.

**Comment by roryabraham** [comment]
I'm going on parental leave, reassigning this back to @cristipaval

**Comment by s77rt** [comment]
@tienifr Thanks for the proposal. Can we have the composer take full height without the use of `HEADER_HEIGHT`?

**Comment by s77rt** [comment]
Actually the use of HEADER_HEIGHT is ok. Can you update your proposal to use `reportActionsDrafts_` to store the composer size status

**Comment by NJ-2020** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Add full screen expand button when editing a message
### What is the root cause of that problem?
New feature
### What changes do you think we should make in order to solve the problem?
- We need to add the expand button inside edit message input composer similar what we do inside AttachmentPickerWithMenuItems.tsx when editing the message
https://github.com/Expensify/App/blob/76c5c81542e598c147f57cf149122f4e124dfe10/src/pages/home/report/ReportActionItemMessageEdit.tsx#L479-L481
```tsx
<View style={[styles.dFlex, styles.flexColumn, isFullComposerAvailable || isComposerFullSize ? styles.justifyContentBetween : styles.justifyContentCenter, styles.mb1]}>
    {isComposerFullSize && (
        <Tooltip text={translate('reportActionCompose.collapse')}>
            <PressableWithFeedback
                onPress={(e) => {
                    e?.preventDefault();
                    Report.setIsComposerFullSize(action.reportActionID, false, true);
                }}
                // Keep focus on the composer when Collapse button is clicked.
                onMouseDown={(e) => e.preventDefault()}
                style={styles.composerSizeButton}
                disabled={isBlockedFromConcierge}
                role={CONST.ROLE.BUTTON}
                accessibilityLabel={translate('reportActionCompose.collapse')}
            >
                <Icon
                    fill={theme.icon}
                    src={Expensicons.Collapse}
                />
            </PressableWithFeedback>
        </Tooltip>
    )}
    {!isComposerFullSize && isFullComposerAvailable && (
        <Tooltip text={translate('reportActionCompose.expand')}>
            <PressableWithFeedback
                onPress={(e) => {
                    e?.preventDefault();
                    Report.setIsComposerFullSize(action.reportActionID, true, true);
                }}
                // Keep focus on the composer when Expand button is clicked.
                onMouseDown={(e) => e.preventDefault()}
                style={styles.composerSizeButton}
                disabled={isBlockedFromConcierge}
                role={CONST.ROLE.BUTTON}
                accessibilityLabel={translate('reportActionCompose.expand')}
            >
                <Icon
                    fill={theme.icon}
                    src={Expensicons.Expand}
                />
            </PressableWithFeedback>
        </Tooltip>
    )}
    <Tooltip text={translate('common.cancel')}>
        <PressableWithFeedback
            onPress={deleteDraft}
            style={styles.composerSizeButton}
            role={CONST.ROLE.BUTTON}
            accessibilityLabel={translate('common.close')}
            // disable dimming
            hoverDimmingValue={1}
            pressDimmingValue={1}
            // Keep focus on the composer when cancel button is clicked.
            onMouseDown={(e) => e.preventDefault()}
        >
            <Icon
                fill={theme.icon}
                src={Expensicons.Close}
            />
        </PressableWithFeedback>
    </Tooltip>
</View>
```
- We will use the `reportActionsDrafts.isEditComposerFullSize` to determine whether the edit composer is full size or not and 
```tsx
const originalReportID = useMemo(() => ReportUtils.getOriginalReportID(reportID, action) || '-1', [reportID, action]);
const [draftActionReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}`, {
    selector: (draftMessagesForReport) => draftMessagesForReport?.[action.reportActionID],
});
const isEditComposerFullSize = useMemo(() => draftActionReport?.isEditComposerFullSize ?? false, [draftActionReport]);
```
- And we need to update some styles and pass some parameters to the composer similar like what we do for sending message
```tsx
<Composer
    ...
    setIsFullComposerAvailable={setIsFullComposerAvailable} // when expand button should be show
    isFullComposerAvailable={isFullComposerAvailable} // 
    isComposerFullSize={isEditComposerFullSize} // is the composer in full size (expand) mode
/>
```
- And we need to set the overflow to hidden and flex1 (for multiple messages we can also use h100) and scroll disabled for ReportActionsList InvertedList and pass some other params when the edit composer is in full size
```tsx
setIsFullComposerAvailable={setIsFullComposerAvailable}
contentContainerStyle={[isEditComposerFullSize && styles.overflowHidden, ...]}
```
- And when expanding the composer we should set the edit composer report action item and others height to 100% i.e `isEditComposerFullSize && styles.chatItemFullComposeRow`

 And we also need some changes inside CellRendererComponent which we will use the position absolute and add position relative to the parent InvertedFlatList when editing the message composer in full size, because if we do not use position absolute it the composer do not cover the remaining messages and I've tried adding the position absolute to ReportActionItem, ReportActionItemMessageEdit and others and it didn't worked
```tsx
function CellRendererComponent({reportID, item: action, style, index, ...props}: CellRendererComponentProps) {
    const originalReportID = useMemo(() => ReportUtils.getOriginalReportID(reportID ?? '-1', action) || '-1', [reportID, action]);
    const [draftActionReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}`, {
        selector: (draftMessagesForReport) => draftMessagesForReport?.[action.reportActionID],
    });

    const isEditComposerFullSize = useMemo(() => draftActionReport?.isEditComposerFullSize ?? false, [draftActionReport]);
    return (
        <View
            {...props}
            style={[
                style,
                isEditComposerFullSize ? {position: 'absolute', height: '100%', width: '100%'} : {zIndex: -index, position: 'relative'},
            ]}
        />
    );
}
```
Alternatively we can use style height inside ReportActionItemMessageEdit `height: windowHeight - HEADER_HEIGHT - 16` but it's not always good to use height like when the ReportFooter is not showing and we have some messages i.e 15 the composer showing like this:
![Screenshot 2024-10-04 at 01 18 36](https://github.com/user-attachments/assets/82e8edcb-e3b1-4b89-8f17-e3fb1bd5ea78)
You can see the gap between the main chat and the workspace header

The rest of the changes can be found in this [test branch](https://github.com/NJ-2020/Expensify/tree/test/15596)
And also we make improvement later on the PR like for CellComponentRenderer and other components

**Here's the [test branch](https://github.com/NJ-2020/Expensify/tree/test/15596)**
### Result

https://github.com/user-attachments/assets/9c28dc78-028b-4780-a331-b9dbbb3e3dbc

### What alternative solutions did you explore? (Optional)

**Comment by NJ-2020** [comment]
### [Proposal Updated](https://github.com/Expensify/App/issues/15596#issuecomment-2393159443)

**Comment by NJ-2020** [comment]
### [Proposal Updated](https://github.com/Expensify/App/issues/15596#issuecomment-2393159443)
- Code replacement for point no 1

**Comment by s77rt** [comment]
@NJ-2020 Thanks for the proposal.

> But it's not always good to use height like the above solution because when the ReportFooter is not showing and we have some messages i.e 15 the composer showing like this

I'm not sure what you meant here. The screenshot looks good to me. Can you please elaborate on what's wrong?

**Comment by s77rt** [comment]
@tienifr

> Hide main composer if isFullSize (from reportActionsDrafts_) is true

This will cause a confusing UI if the message that is in full size mode is not visible (not in view). Can we maybe hide the composer only if the edit message is focused? (or always keep the main composer visible)

**Comment by NJ-2020** [comment]
@s77rt I tested the tienifr solution again using the height style i.e `height: windowHeight - HEADER_HEIGHT - 16` 
And it's showing like this with only 2 messages
![Screenshot 2024-10-05 at 05 18 26](https://github.com/user-attachments/assets/4c6981c9-1ff1-4d28-ab32-4d1c3b64066b)
You can see the gap between the main chat and the header
And when i tested again with many messages i.e 15 and click one of the middle of the message, i see this:
![Screenshot 2024-10-05 at 05 18 53](https://github.com/user-attachments/assets/456069dc-b7d4-4f66-b856-96351d8eab5e)

So as you can see if we are using the fixed height style value inside ReportActionItemMessageEdit.tsx file it's showing inconsistency and invalid result

Please let me know if i miss something here
And you can also test the tienifr test branch solution and make sure to pull the latest commit from main

**Comment by NJ-2020** [comment]
### [Proposal Updated](https://github.com/Expensify/App/issues/15596#issuecomment-2393159443)
- Add new point for Hiding the main composer
- Some other changes

**Comment by s77rt** [comment]
@NJ-2020 Can you please elaborate on the styling i.e. which components will have absolute position? Also I have same concern regarding an edit message in full size but not focused (https://github.com/Expensify/App/issues/15596#issuecomment-2395001634) how will that effect the behaviour i.e. can we still use scroll thought the messages normally?

**Comment by NJ-2020** [comment]
@s77rt We can have a position absolute inside CellRendererComponent because I've tried adding the position absolute inside ReportActionItem, ReportActionItemMessageEdit and others but they didn't work
```tsx
function CellRendererComponent({reportID, item: action, style, index, ...props}: CellRendererComponentProps) {
    const originalReportID = useMemo(() => ReportUtils.getOriginalReportID(reportID ?? '-1', action) || '-1', [reportID, action]);
    const [draftActionReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}`, {
        selector: (draftMessagesForReport) => draftMessagesForReport?.[action.reportActionID],
    });

    const isEditComposerFullSize = useMemo(() => draftActionReport?.isEditComposerFullSize ?? false, [draftActionReport]);
    return (
        <View
            {...props}
            style={[
                style,
                isEditComposerFullSize ? {position: 'absolute', height: '100%', width: '100%'} : {zIndex: -index, position: 'relative'},
            ]}
        />
    );
}
```

But if we don't want to add the position absolute inside the CellRendererComponent because it can cause other lists too, we can create new seperate component i.e CellRendererReportActionMessageComponent.tsx (we will use similar code like the above)and we will use this when rendering the report action message i.e ReportActionsList InvertedList and do some other changes

And regarding when the edit message composer is in full size but not focused, we cannot scroll the messages and it's still cover the entire main chat same like what we do when sending messages
```tsx
contentContainerStyle={[isEditComposerFullSize && styles.overflowHidden, ...]}
scrollDisabled={isEditComposerFullSize ? true : false}
```

**Comment by s77rt** [comment]
@NJ-2020
> when the edit message composer is in full size but not focused, we cannot scroll the messages and it's still cover the entire main chat

I think this could backfire, consider the case where you have many messages: message1...message200 and message50 is in edit mode, and you are currently viewing the last 100 messages, now if you keep scrolling up, you will eventually load message50 and it will suddenly cover the screen which could be confusing

I have more concerns regarding the absolute position:
1. Won't this effect the unread marker logic? If you get a new message, will it still be marked as new? and will the message auto-scroll?
2. What happens if we try to TAB navigate?

**Comment by s77rt** [comment]
@tienifr I don't think we have a clear expected results but the edit composer shouldn't just suddenly block the user from scrolling.

**Comment by s77rt** [comment]
If we don't store the fullsize status of edit messages then I think we can avoid the problem, but this will also remove a feature (not saving to onyx). I do think we can discuss the removal of that feature if it will help get this shipped faster and easier

**Comment by s77rt** [comment]
cc @cristipaval ^ Does the removal of saving the fullsize statue in onyx sounds like something we can go with?

Pros: Make this feature easier to implement
Cons: If you switch an edit composer to full size (press expand button) then refresh the page, the edit composer will not be in full size

**Comment by s77rt** [comment]
@NJ-2020 Were you able to verify the points mentioned in https://github.com/Expensify/App/issues/15596#issuecomment-2395437129

**Comment by s77rt** [comment]
@tienifr In terms of UX I think you shouldn't be able scroll until you shrink back the composer. I have not confirmed the mentioned concerns yet. Did you notice any strange behaviour?

**Comment by NJ-2020** [comment]
@s77rt Sorry for the delay
> Were you able to verify the points

Yes
> think this could backfire, consider the case where you have many messages: message1...message200 and message50 is in edit mode, and you are currently viewing the last 100 messages, now if you keep scrolling up, you will eventually load message50 and it will suddenly cover the screen which could be confusing

The edit composer will only cover the main chat if it's in full size mode and inside the current screen - meaning if the edit composer inside reportID 10 then we go to another chat, the edit composer will not cover the main chat and the scroll will not be disabled

So when we go back to the current screen the edit composer will only cover the main chat and prevent the user scrolling from the messages when the edit composer is in full size inside the current screen

> Won't this effect the unread marker logic? If you get a new message, will it still be marked as new? and will the view auto-scroll?

I am still finding out to solve this issue

When the edit composer is in full size then if we collapse the edit composer it will automatically scroll to the bottom which will dismiss the unread marker logic, it cause by (`flex1` or `h100`) and the overflowHIdden

Here's the current style we use to disable the scroll which will dismiss the unread marker logic
```tsx
const contentContainerStyle: StyleProp<ViewStyle> = useMemo(
    () => [
        isEditComposerFullSize && [styles.h100],
        styles.chatContentScrollView,
        isLoadingNewerReportActions && canShowHeader ? styles.chatContentScrollViewWithHeaderLoader : {},
    ],
    [isLoadingNewerReportActions, styles.chatContentScrollView, styles.chatContentScrollViewWithHeaderLoader, canShowHeader, isEditComposerFullSize],
);
```
When we add those styles and the expand button is clicked, the height of the container becoming smaller which is the height of the current screen, meaning if we have a browser height of 1000px then the container height will be around 1000px - header - others and will be automatically scroll to the bottom

I've tried removing the overflowHidden, flex adding overflow auto, forcing the child CellRendererComponent to have a full height position fixed covering the main chat, and some other way, but still they didn't work

<hr />

**Solution**

We can invoke the ReportActionItemMessageEdit component within the ReportFooter or ReportScreen, then we can call this component near the ReportActionCompose component and pass a full-size mode to true (reportActionDraft.isEditComposerFullSize) when the expand button is clicked. 

I can see that the ReportActionCompose covers the main chat by adding h100 and flex1 and some other components, and it does not affect the scroll issue (unread marker logic)

```tsx
const [reportActionDrafts] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${report.reportID}`);
const [fullSizeEditMessageDraftID] = useMemo(
    () => Object.keys(reportActionDrafts ?? {}).filter((reportActionDraftID) => reportActionDrafts?.[reportActionDraftID].isEditComposerFullSize === true),
    [reportActionDrafts],
);
const [reportActions] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report.reportID}`);
const reportAction = useMemo(() => reportActions?.[fullSizeEditMessageDraftID ?? '-1'], [reportActions]);
const fullSizeEditMessageDraft = useMemo(() => reportActionDrafts?.[fullSizeEditMessageDraftID], [reportActionDrafts]);
...
{!shouldHideComposer && (!!shouldShowComposeInput || !shouldUseNarrowLayout) && (
    <View style={[chatFooterStyles, (isComposerFullSize || fullSizeEditMessageDraft?.isEditComposerFullSize) && styles.chatFooterFullCompose]}>
        <SwipeableView onSwipeDown={Keyboard.dismiss}>
            {!fullSizeEditMessageDraft?.isEditComposerFullSize && (
                <ReportActionCompose
                    onSubmit={onSubmitComment}
                    onComposerFocus={onComposerFocus}
                    onComposerBlur={onComposerBlur}
                    reportID={report.reportID}
                    report={report}
                    lastReportAction={lastReportAction}
                    pendingAction={pendingAction}
                    isComposerFullSize={isComposerFullSize}
                    isReportReadyForDisplay={isReportReadyForDisplay}
                    shouldShowEducationalTooltip={didScreenTransitionEnd && shouldShowEducationalTooltip}
                />
            )}
            {fullSizeEditMessageDraft?.isEditComposerFullSize && reportAction && (
                <ReportActionItemMessageEdit
                    action={reportAction}
                    draftMessage={fullSizeEditMessageDraft?.message ?? ''}
                    reportID={report.reportID}
                    policyID={report?.policyID}
                    index={0}
                    ref={textEditInputRef}
                    shouldDisableEmojiPicker={
                        (ReportUtils.chatIncludesConcierge(report) && User.isBlockedFromConcierge(blockedFromConcierge)) ||
                        ReportUtils.isArchivedRoom(report, reportNameValuePairs)
                    }
                    isGroupPolicyReport={!!report?.policyID && report.policyID !== CONST.POLICY.ID_FAKE}
                    isFullSize={fullSizeEditMessageDraft.isEditComposerFullSize}
                />
            )}
        </SwipeableView>
    </View>
)}
```
The idea is to implement the edit composer inside the ReportFooter or ReportScreen because it was so easy to cover the main chat and disable the scroll without needing any changes inside the ReportActionsList and unread messages will still be marked as new and will not automatically scroll to the bottom

### Result

https://github.com/user-attachments/assets/24588d30-22a1-4386-ad06-12ff4dbbac91

Please let me know what do you think

**Comment by s77rt** [comment]
@NJ-2020 Regarding the first point, I think we are not on the same page. I'm not referring to the change of reports. I'm referring to the same report. Consider a report that has 200 messages and message50 is in edit mode (flagged from onyx). Initially you will load latest messages e.g. message200 to message150. Then if you keep scrolling up, you will eventually load message50 and it will suddenly cover the entire chat.

Regarding the second point:

> The idea is to implement the edit composer inside the ReportFooter or ReportScreen

Does that mean we will have 2 edit composers? (one for expanded edit and the other for normal edit)

**Comment by s77rt** [comment]
@tienifr 

> if message50 is in edit mode, that means it already loaded, right?

Not necessary, if we save the edit status in onyx then message50 will always be in edit mode just not yet rendered

**Comment by NJ-2020** [comment]
> Does that mean we will have 2 edit composers? (one for expanded edit and the other for normal edit)

Yes we will have 2 edit composers, one inside the ReportActionItem.tsx and the second one inside the ReportFooter

But we will show the second one only if the the expand button is clicked meaning `isEditComposerFullSize` is true, and the second edit composer will cover the main chat and prevent the user from scrolling

**Comment by s77rt** [comment]
@NJ-2020 

> https://github.com/Expensify/App/issues/15596#issuecomment-2411199120

Resetting the flag kinda contradicts the purpose of saving it in onyx in the first place. 

> Yes we will have 2 edit composers

I think this will cause sync problems:
1. If you quickly type then click the expand button, the text could have not get saved into onyx yet, making the new expanded composer load with outdated text
2. The cursor position won't be in sync

It's probably best to keep one composer. Maybe we can "teleport" it using createPortal?

**Comment by NJ-2020** [comment]
@s77rt 
> It's probably best to keep one composer. Maybe we can "teleport" it using createPortal?

Yes we can, now we will only have 1 edit composer inside ReportActionItem
```tsx
function ReportActionItemMessageEditWrapper({children, isFullSize}: ReportActionItemMessageEditWrapperProps) {
    const reportFooter = document.querySelector('#reportFooterComposer');
    return isFullSize && reportFooter ? ReactDOM.createPortal(children, reportFooter) : children;
}
return (
    <ReportActionItemMessageEditWrapper isFullSize={isFullSize}>
        <View
            style={[styles.chatItemMessage, styles.flexRow, isFullSize && [styles.chatItemFullComposeRow, styles.pb5]]}
            ref={containerRef}
        >
            ...
```
We will need to add an id to the report footer and we will hide the main composer if `isEditComposerFullSize` is true

### Result

https://github.com/user-attachments/assets/091b9435-53f7-4045-aba0-aacfc71c2944

**Comment by s77rt** [comment]
@NJ-2020 Actually I think this makes no difference than having two components. The element is still getting removed from the dom and the sync issues are still present.

**Comment by s77rt** [comment]
Still waiting for proposals... (even for proposals that just make this work without Onyx)

**Comment by Anaslancer** [comment]
@s77rt Can we make a new variable for each edit box?
`const [fullSizeEditComposer] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_IS_EDIT_COMPOSER_FULL_SIZE}`, {initialValue: reportActionID: -1});`
So we can figure out which report is being a full size.

**Comment by s77rt** [comment]
@Anaslancer For now the storage and the state are not much of a concern. We still need to find a way to expand the edit composer. Were you able to look into that?

**Comment by Anaslancer** [comment]
@s77rt Thanks! Yes, I will find more with adding new variable to storage.

**Comment by s77rt** [comment]
@Anaslancer I don't think I understood your question. We have two composers, the main one for sending messages (which already support the expand option) and the edit composer which is rendered inline as a drop replacement for the report action itself.

**Comment by Anaslancer** [comment]
@s77rt We don't need to merge 2 composers, right?

**Comment by s77rt** [comment]
@Anaslancer That's right

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
Still no proposals yet

**Comment by QichenZhu** [comment]
@s77rt Could you summarize the reasons the previous proposals were not accepted, as this discussion is quite long?

> Still waiting for proposals... (even for proposals that just make this work without Onyx)

What other key characteristics should new proposals have compared to the old ones?

Thanks!

**Comment by s77rt** [comment]
@QichenZhu When clicking on the expand button the following are expected:
- Message view takes full height
- The focus and the cursor position are not lost
- The unread indicator functions normally i.e. if someone sends you a message it should be marked as unread after you close the edit composer
- ~Save state in onyx~ I don't think we should focus on this one as it's the reason we didn't move forward so far

**Comment by s77rt** [comment]
cc @tienifr I think your proposal would work too if we omit the save in onyx feature, would you be able to double check your proposal as well

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
Same ^

**Comment by dannymcclain** [comment]
I kinda don't see why we wouldn't. Curious for other design team thoughts but I feel like the "composer" should basically always be the same.

**Comment by shawnborton** [comment]
I don't feel strongly really. I can see the argument for consistency but I can also see where this one might be complicated to pull off correctly since the message that is being edited might not necessarily be docked to the bottom of the view or anything.

**Comment by cristipaval** [comment]
Thanks, @sonialiap and @Expensify/design! I'll close the issue for now and come back to it if the users want it, given the ton of other priorities we have atm.
